zboxcore/sdk/chunked_upload.go

```diff
@@ -527,38 +509,69 @@ func (su *ChunkedUpload) processUpload(chunkStartIndex, chunkEndIndex int, fileS
 			return err
 		}
 
-		go func(b *ChunkedUploadBlobber, buf *bytes.Buffer, form ChunkedUploadFormMetadata) {
+		go func(idx uint64, b *ChunkedUploadBlobber, buf *bytes.Buffer, form ChunkedUploadFormMetadata) {
 			err := b.sendUploadRequest(ctx, su, chunkEndIndex, isFinal, encryptedKey, buf, form)
 
 			// channel is not closed
 			if ctx.Err() == nil {
-				wait <- err
+				wait <- UploadError{
+					Error:      err,
+					BlobberIdx: pos,
+				}
 			}
-		}(blobber, body, formData)
+		}(pos, blobber, body, formData)
 	}
-	var err error
+	var err UploadError
 	for i := 0; i < num; i++ {
 		err = <-wait
-		if err != nil {
-			return err
+		if err.Error != nil {
+			logger.Logger.Error("Upload: ", err.Error)
+			//stop to upload new chunks to failed blobber
+			su.uploadMask = su.uploadMask.Sub64(err.BlobberIdx)
 		}
 	}
 
-	return nil
+	if su.consensus.isConsensusOk() {
+		return nil
+	}
+
+	// all of blobber are failed. it should be rejected by rule
+	if su.consensus.getConsensus() == 0 {
+		return err.Error
+	}
+
+	errConsensus := fmt.Errorf("Upload failed: Consensus_rate:%f, expected:%f", su.consensus.getConsensusRate(), su.consensus.getConsensusRequiredForOk())
+	if su.statusCallback != nil {
+		su.statusCallback.Error(su.allocationObj.ID, su.fileMeta.Path, OpUpload, errConsensus)
+	}
+
+	return errConsensus
 
 }
 
 // processCommit commit shard upload on its blobber
 func (su *ChunkedUpload) processCommit() error {
+	shouldUnlock := true
+	err := su.writeMarkerMutex.Lock(context.TODO(), su.progress.ConnectionID)
+	defer func() {
```

Comment:
Check err before the defer as if we lock failed, it is unnecessary to call unlock

Reply:
we still need try to unlock all of them. because some of blobbers might be locked. eg 1-3 are locked, but fails to lock 4th blobber.

Reply:
and in WriteMarkerMutex, it will skip to unlock if a blobber is not locked.
```
		blobberUrl := builder.String()
		// only release lock on locked blobbers
		if m.lockedBlobbers != nil {
			if m.lockedBlobbers[b.ID] {
				urls = append(urls, blobberUrl)
			}
		} else { // Lock is not called here, try to release all blobbers
			urls = append(urls, blobberUrl)
		}
```